src/main/telemetry/esc_telemetry.c

```diff
@@ -0,0 +1,319 @@
+/*
+KISS ESC TELEMETRY PROTOCOL
+---------------------------
+
+One transmission will have 10 times 8-bit bytes sent with 115200 baud and 3.6V.
+
+Byte 0: Temperature
+Byte 1: Voltage high byte
+Byte 2: Voltage low byte
+Byte 3: Current high byte
+Byte 4: Current low byte
+Byte 5: Consumption high byte
+Byte 6: Consumption low byte
+Byte 7: Rpm high byte
+Byte 8: Rpm low byte
+Byte 9: 8-bit CRC
+
+*/
+
+/*
+DEBUG INFORMATION
+-----------------
+
+set debug_mode = DEBUG_ESC_TELEMETRY in cli
+
+0: current motor index requested
+1: number of timeouts
+2: voltage
+3: current
+*/
+
+#ifdef USE_DSHOT
+
+typedef struct {
+    bool skipped;
+    int16_t temperature;
+    int16_t voltage;
+    int16_t current;
+    int16_t consumption;
+    int16_t rpm;
+} esc_telemetry_t;
+
+typedef enum {
+    ESC_TLM_FRAME_PENDING = 1 << 0,     // 1
+    ESC_TLM_FRAME_COMPLETE = 1 << 1     // 2
+} escTlmFrameState_t;
+
+typedef enum {
+    ESC_TLM_TRIGGER_WAIT = 0,
+    ESC_TLM_TRIGGER_READY = 1 << 0,     // 1
+    ESC_TLM_TRIGGER_PENDING = 1 << 1,   // 2
+} escTlmTriggerState_t;
+
+#define ESC_TLM_BAUDRATE 115200
+#define ESC_TLM_BUFFSIZE 10
+#define ESC_BOOTTIME 5000               // 5 seconds
+#define ESC_REQUEST_TIMEOUT 100         // 100 ms (data transfer takes only 900us)
+
+static bool tlmFrameDone = false;
+static uint8_t tlm[ESC_TLM_BUFFSIZE] = { 0, };
+static uint8_t tlmFramePosition = 0;
+static serialPort_t *escTelemetryPort = NULL;
+static esc_telemetry_t escTelemetryData[MAX_SUPPORTED_MOTORS];
+static uint32_t escTriggerTimestamp = -1;
+static uint32_t escTriggerLastTimestamp = -1;
+static uint8_t enabledMotorCount;
+static uint8_t timeoutRetryCount = 0;
+
+static uint8_t escTelemetryMotor = 0;      // motor index
+static bool escTelemetryEnabled = false;
+static escTlmTriggerState_t escTelemetryTriggerState = ESC_TLM_TRIGGER_WAIT;
+
+static int16_t escVbat = 0;
+static int16_t escCurrent = 0;
+static int16_t escConsumption = 0;
+
+static void escTelemetryDataReceive(uint16_t c);
+static uint8_t update_crc8(uint8_t crc, uint8_t crc_seed);
+static uint8_t get_crc8(uint8_t *Buf, uint8_t BufLen);
+static void selectNextMotor(void);
+
+bool isEscTelemetryActive(void)
+{
+    return escTelemetryEnabled;
+}
+
+int16_t getEscTelemetryVbat(void)
+{
+    return escVbat / 10;
+}
+
+int16_t getEscTelemetryCurrent(void)
+{
+    return escCurrent;
+}
+
+int16_t getEscTelemetryConsumption(void)
+{
+    return escConsumption;
+}
+
+bool escTelemetryInit(void)
+{
+    serialPortConfig_t *portConfig = findSerialPortConfig(FUNCTION_TELEMETRY_ESC);
+    if (!portConfig) {
+        return false;
+    }
+
+    portOptions_t options = (SERIAL_NOT_INVERTED);
+
+    // Initialize serial port
+    escTelemetryPort = openSerialPort(portConfig->identifier, FUNCTION_TELEMETRY_ESC, escTelemetryDataReceive, ESC_TLM_BAUDRATE, MODE_RX, options);
+
+    if (escTelemetryPort) {
+        escTelemetryEnabled = true;
+        masterConfig.batteryConfig.currentMeterType = CURRENT_SENSOR_ESC;
```

Comment:
I think that is the right way to do it. Once we have MSP (or MSP's successor) commands for this functionality, we can always add code to the configurator to set current and voltage sensor to ESC when the ESC_TELEMETRY feature is enabled.